File: 121_BestTimeToBuyAndSellStock/max_profit.cpp
// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/?envType=study-plan-v2&envId=top-interview-150

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // set initial buy price to day 1 price
        int buy_price = prices[0];
        // initialize max profit to 0
        int max_profit = 0;

        // iterate through each price
        for(int i = 0; i < prices.size(); i++)
        {
            // check if we have found a lower potential buy price
            if(prices[i] < buy_price)
            {
                // update lowest possible buy price for following days
                buy_price = prices[i];
            } 
            // otherwise, check profit for selling on this day
            else if(prices[i] - buy_price > max_profit)
            {
                // update max potential profit
                max_profit = prices[i] - buy_price;
            }
        }

        return max_profit;
    }
};

int main() {
    Solution solution;

    // Input vector size
    cout << "Enter the number of days (size of prices array): ";
    int n;
    cin >> n;

    // Input prices array
    vector<int> prices(n);
    cout << "Enter the prices: ";
    for (int i = 0; i < n; i++) {
        cin >> prices[i];
    }

    // Call maxProfit function
    int result = solution.maxProfit(prices);

    // Output the result
    cout << "Maximum profit: " << result << endl;

    return 0;
}

File: 122_BestTimeToBuyAndSellStockII/max_profit.py
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # Initialize profit to 0
        profit = 0

        # Iterate through all prices
        for i, price in enumerate(prices):
            # Check that we are past day one
            if i > 0:
                # Check if a profit could have been made between today and yesterday
                if price > prices[i - 1]:
                    # Add profit
                    profit += price - prices[i - 1]

        return profit

# Test cases
if __name__ == "__main__":
    solution = Solution()
    
    # Example test cases
    test_cases = [
        ([7,1,5,3,6,4], 7),
        ([1,2,3,4,5], 4),
        ([7,6,4,3,1], 0),
    ]

    for prices, expected in test_cases:
        result = solution.maxProfit(prices)
        print(f"Prices: {prices} => Max Profit: {result} (Expected: {expected})")
